### PR TITLE
fix(neighbors): size batched cluster_tile buffers by density

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Update batched cluster-tile sizing so Warp, JAX, and Torch use a shared per-system density estimate for `max_tiles_per_group`, and expose the estimator as public API. This prevents dense high-cutoff batched systems from reusing an undersized fixed compact tile buffer while preserving the historical floor for small batches.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

N/A

## Changes Made

- Add shared `estimate_batch_max_tiles_per_group` support in the core cluster-tile launchers with validation for batch pointers and per-system cell volumes.
- Expose the batched max-tile estimator through the JAX and Torch neighbor APIs, and route automatic batched cluster-tile allocation through the shared estimator.
- Document the new estimator APIs and add JAX, Torch, and core sizing tests for dense sizing, concrete-input validation, small-batch floor behavior, and manual `max_tiles_per_group` bypass.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Staged diff whitespace check passed. Unit tests and lint were not run while preparing these PR materials.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

None.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/jax/neighbors/batch_cluster_tile.py [hunk 1]
```diff
@@ -92,23 +93,95 @@ def _batch_tile_buffer_max_tiles_per_group(
     vols = _concrete_cell_batch_volumes(cell_batch)
     if isinstance(positions, jax.core.Tracer) or vols is None or not cutoff_concrete:
         return max(max_ng, 1)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/batch_cluster_tile.py]: Fixed in ee2c5e3a8031771c1fdff468a7c437ff30a4b15f.

File: nvalchemiops/torch/neighbors/batch_cluster_tile.py [hunk 2]
```diff
@@ -115,48 +119,42 @@ def _(
...
-            2.0 * (4.0 / 3.0) * torch.pi * radii * radii * radii / cluster_volumes,
-        )
-        .to(torch.int64)
-        .clamp_min(floor)
+    Returns
+    -------
+    int
+        Shared ``max_tiles_per_group`` for the batched compact tile buffer.
+    """
+    volumes = torch.linalg.det(cell_batch).abs().view(-1)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/neighbors/batch_cluster_tile.py]: Fixed in ee2c5e3a8031771c1fdff468a7c437ff30a4b15f.

File: nvalchemiops/jax/neighbors/batch_cluster_tile.py [hunk 3]
```diff
@@ -92,23 +93,95 @@ def _batch_tile_buffer_max_tiles_per_group(
...
+            arr = np.asarray(cell_batch)
+        except Exception as exc:
+            raise ValueError(
+                "cell_batch must be concrete to estimate batch max_tiles_per_group"
+            ) from exc
+        if arr.ndim != 3 or arr.shape[1:] != (3, 3):
+            raise ValueError("cell_batch must have shape (num_systems, 3, 3)")
+        raise ValueError(
+            "cell_batch must be concrete to estimate batch max_tiles_per_group"
+        )
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/batch_cluster_tile.py]: Fixed in ee2c5e3a8031771c1fdff468a7c437ff30a4b15f.

File: nvalchemiops/neighbors/cluster_tile/launchers.py [hunk 4]
```diff
@@ -105,6 +107,72 @@ def estimate_max_tiles_per_group(
...
 
+def _host_int_float_list(values) -> list[int | float]:
+    if hasattr(values, "detach"):
+        return values.detach().cpu().tolist()
+    if hasattr(values, "tolist"):
+        return values.tolist()
+    return list(values)
+
+
+def estimate_batch_max_tiles_per_group(
```
Comment (laserkelvin) [inline nvalchemiops/neighbors/cluster_tile/launchers.py:119]: Missing type hints on `batch_ptr` and `cell_volumes`

File: test/neighbors/bindings/jax/test_batch_cluster_tile.py [hunk 5]
```diff
@@ -254,6 +257,97 @@ def test_bad_cell_shape_raises(self):
...
+                    int(stop - start),
+                    cutoff,
+                    float(abs(np.linalg.det(cell))),
+                ),
+            )
+
+        got = estimate_batch_max_tiles_per_group(batch_ptr, cutoff, cell_batch)
+
+        assert got == expected
+        assert got > 256
```
Comment (laserkelvin) [inline test/neighbors/bindings/jax/test_batch_cluster_tile.py:290]: Don't really understand this assertion

File: test/neighbors/bindings/jax/test_cluster_tile.py [hunk 6]
```diff
@@ -741,8 +741,8 @@ class TestJaxClusterTileTileSizing:
     """The JAX tile buffer must geometry-size (concrete) / fall back (traced).
 
     Guards the fix that stopped the JAX default/auto-select path from silently
-    undercounting dense high-cutoff systems (fixed ``max_tiles_per_group=256``
-    overflowed; e.g. ammonia ngroup=1024 at 20/25 A).
+    undercounting dense high-cutoff systems where a fixed
+    ``max_tiles_per_group=256`` cannot cover all neighboring row groups.
```
Comment (laserkelvin) [inline test/neighbors/bindings/jax/test_cluster_tile.py:745]: Was this actually modified?

File: test/neighbors/bindings/torch/test_batch_cluster_tile.py [hunk 7]
```diff
@@ -217,6 +223,68 @@ def test_batch_cluster_tile_max_tiles_per_group_vectorized_helper():
     assert got > 256
```
Comment (laserkelvin) [inline test/neighbors/bindings/torch/test_batch_cluster_tile.py:231]: Same as the other test - sorry if I'm missing something

**Review by laserkelvin — CHANGES_REQUESTED** [review]
Generally fine with me, some minor comments

File: test/neighbors/bindings/jax/test_cluster_tile.py (hunk 6 above)
Comment (zubatyuk) [reply test/neighbors/bindings/jax/test_cluster_tile.py:745]: This is docstring cleanup.

File: test/neighbors/bindings/jax/test_batch_cluster_tile.py (hunk 5 above)
Comment (zubatyuk) [reply test/neighbors/bindings/jax/test_batch_cluster_tile.py:290]: Batched estimator starts at the floor and only exceeds it if at least one system’s scalar estimate exceeds 256.

File: nvalchemiops/neighbors/cluster_tile/launchers.py (hunk 4 above)
Comment (zubatyuk) [reply nvalchemiops/neighbors/cluster_tile/launchers.py:119]: Fixed in b0acfe82b1d79afc8cd0ca73c8bcad47592e0fd8.

File: nvalchemiops/neighbors/cluster_tile/launchers.py [hunk 8]
```diff
@@ -105,6 +108,72 @@ def estimate_max_tiles_per_group(
...
+    """Estimate batched ``max_tiles_per_group`` from per-system density.
+
+    The compact batched tile buffer is ``ngroup_total * min(ngroup_total,
+    max_tiles_per_group)``, so the shared ``max_tiles_per_group`` must cover
+    the densest system.  Returns the maximum per-system estimate from
+    :func:`estimate_max_tiles_per_group`, floored at ``floor``.
+
+    Parameters
+    ----------
+    batch_ptr : sequence of int
```
Comment (laserkelvin) [inline nvalchemiops/neighbors/cluster_tile/launchers.py:136]: Align with `Iterable`

**Review by laserkelvin — APPROVED** [review]
Approved, but just fix the one docstring and then we can merge
